### PR TITLE
Remove UnificationError

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes https://github.com/kframework/kore/issues/1767

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/test/Test/Kore/Step/RewriteStep.hs [hunk 1]
```diff
@@ -445,25 +435,12 @@ test_applyRewriteRule_ =
...
-    -- vs
-    -- sigma(a, i(b)) with substitution b=a
-    , testCase "non-function substitution error" $ do
-        let expect = Left $ unsupportedPatterns
-                "Unknown unification case."
-                (Mock.plain10 (mkElemVar Mock.y))
-                (mkElemVar Mock.x)
-            initial = pure $
-                Mock.sigma (mkElemVar Mock.x) (Mock.plain10 (mkElemVar Mock.y))
-        actual <- applyRewriteRuleParallel_ initial axiomSigmaId
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Step/RewriteStep.hs:458]: This test will now throw the following error:
```
Expected function-like term, but found:
    /* Created: Test.Kore.Step.RewriteStep.sigma */
    /* Sfa */
    sigma{}(
        /* Fl Fn D Sfa */ Configx:testSort{},
        /* Created: Test.Kore.Step.RewriteStep.plain10 */
        /* Sfa */
        plain10{}(/* Fl Fn D Sfa */ Configy:testSort{})
    )
CallStack (from HasCallStack):
  error, called at src/Kore/Step/Step.hs:195:21 in kore-0.23.0.0-Kq7aVMskyU0HcW2wehu1H0:Kore.Step.Step
```

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 2]
```diff
@@ -137,14 +134,12 @@ termUnification notSimplifier =
...
-                throwUnificationError
-                    (unsupportedPatterns
-                        "Unknown unification case."
-                        pat1
-                        pat2
-                    )
-        Error.maybeT unsupportedPatternsError pure maybeTermUnification
+        Error.maybeT (unificationPattern pat1 pat2) pure maybeTermUnification
+
+    unificationPattern term1 term2 =
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/AndTerms.hs:139]: Let's call this something like `incompleteUnificationPattern` instead.

File: kore/test/Test/Kore/Unification/Unifier.hs [hunk 3]
```diff
@@ -655,11 +652,9 @@ test_unsupportedConstructs =
         andSimplifyFailure
```
Comment (ttuegel) [inline kore/test/Test/Kore/Unification/Unifier.hs:652]: Is there a difference now between `andSimplifySuccess` and `andSimplifyFailure`?

Comment (ana-pantilie) [reply kore/test/Test/Kore/Unification/Unifier.hs]: You're right, there isn't any. I've made the necessary changes. Thanks!

**Pull request merged**